test/metabase/upload_test.clj

```diff
@@ -1915,6 +1925,35 @@
 
           (io/delete-file file))))))
 
+(defn- one-of?
+  "Builds a predicate which checks whether a value is one of the given values. Needed as =? treats sets as literals."
+  [& args]
+  (let [s (set args)]
+    (fn [value]
+      (contains? s value))))
+
+(deftest replace-from-csv-int-and-float-test
+  (mt/test-drivers (mt/normal-drivers-with-feature :uploads)
+    (testing "Append should handle a mix of int and float-or-int values being appended to an int column"
+      (with-upload-table! [table (create-upload-table!
+                                  :col->upload-type (ordered-map/ordered-map
+                                                     :_mb_row_id ::upload/auto-incrementing-int-pk
+                                                     :number_1 ::upload/int
+                                                     :number_2 ::upload/int)
+                                  :rows [[1, 1]])]
+
+        (let [csv-rows ["number-1, number-2"
+                        "1.0, 1"
+                        "1  , 1.0"]
+              file     (csv-file-with csv-rows (mt/random-name))]
+          (is (some? (replace-csv! {:file file, :table-id (:id table)})))
+          ;; For MySQL the primary key incrementer will be reset, but for Postgres and H2 it will not.
```

Comment:
If we decide to go forward with these driver-specific semantics, we should probably tighten up the assertion to be precise per driver.